consensus/src/marshal/actor.rs

```diff
@@ -436,16 +417,22 @@ impl<
                                 resolver.fetch(Request::<B>::Notarized { view }).await;
                             }
 
-                            // Register waiter
-                            debug!(view, ?commitment, "registering waiter");
-                            match self.waiters.entry(commitment) {
+                            // Register listener
+                            debug!(view, ?commitment, "registering listener");
+                            match self.listeners.entry(commitment) {
                                 Entry::Occupied(mut entry) => {
-                                    entry.get_mut().add_responder(response);
+                                    entry.get_mut().listeners.push(response);
                                 }
                                 Entry::Vacant(entry) => {
                                     let (tx, rx) = oneshot::channel();
                                     buffer.subscribe_prepared(None, commitment, None, tx).await;
-                                    entry.insert(Waiter::new(rx, response));
+                                    let hook = waiters.push(async move {
```

Comment:
I wonder if it makes sense to make a "wrapper" struct that handles the waiters and listeners?

The logic you introduced here looks solid but seems it could be simplified if we introduced one additional util (unless the generics get too bad)?